Update nearest node search for iteration patch update strategy

recuero opened the pull request:
~The iteration update strategy fails to work with initial adaptivity, regardless of the `ghosting_patch_size` parameter, i.e. one can make it very large and still get the message that is not large enough. I am adding a hopefully more informative error message. It'd be good to address the underlying problem in the future, so I am leaving the issue open.~

Trying to address the underlying issue in this PR.

Closes #24801

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1611560/) on 1cd2c09 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24805/site)

This comment will be updated on new commits.

recuero committed:
Ensure neighborhood node list is built on reinit for MOOSE::Iteration (#24801)

MOOSE::Iteration update strategy relies on building neighborhood lists at the
initialization stage. When mesh has initial adaptivity, doing this on initialization
is not enough as evidenced by the fact that neighborhood nodes are not found on the list
of ghosted nodes. If we do this on reinitialization, mesh information appeared to have
been updated with refined elements.

This is intended to enable initial adaptivity with Iteration patch update strategy in
parallel runs.

recuero committed:
Add test for iteration patch update strategy with initial adaptivity.

Input file based on that provided by @BoZeng1997 in #24801.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1615774/) on cc2c3f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24805/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24805/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

recuero committed:
Add node-face test with adaptivity/replicated mesh

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1616731/) on cbe5ab0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0e3d4f</th>
      <th colspan=4>#24805 cbe5ab</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.58%</td>
      <td>85.59%</td>
      <td>+0.01%</td>
      <td>60.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89021</td>
      <td>89028</td>
      <td>+7</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14997</td>
      <td>14992</td>
      <td>-5</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24805/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24805/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24805/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 60.00% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/src/geomsearch/NearestNodeLocator.C:
```diff
@@ -69,8 +70,13 @@ NearestNodeLocator::findNodes()
+    // After a call from system reinit, mesh has been updated with initial adaptivity.
+    // Moose::Iteration relies on data generated for ghosting (i.e. trial_primary_nodes)
+    if (!_first)
+      _reinit_iteration = false;
```
I m not sure about this logic. 
Seems like we:
- findNodes on first
- the second time, we do it again because reinit_iteration started true and is still true
- the third time we dont because reinit_iteration is not true

should we just keep track of the number of iterations instead of these two booleans?

recuero replied:
The purpose is to correctly identify nearest nodes when we do initial adaptivity. To that effect, we need to go through that piece of code upon `reinit()`, in addition to `init()`. 

There may be other ways to address that problem. We could keep track of the number of iterations. However, that code gets called a few times before we attempt to the first nonlinear iteration. If we based the logic on the nonlinear iterations, we would probably be going through that code more often than needed.

All that said, I am not very familiar with how initial adaptivity or adaptivity in general is supposed to work with the calls within `FEProblemBase`, so I'd be happy to change the logic.

GiudGiud replied:
I dont see any logic that captures this:
```
The purpose is to correctly identify nearest nodes when we do initial adaptivity. To that effect, we need to go through that piece of code upon reinit(), in addition to init().
```

recuero replied:
This code gets called a second time on reinit. That's enabled by the reinit flag you commented on above.

GiudGiud replied:
can you set this boolean on reinit() instead?

recuero committed:
Address Guillaume's review.

The flag to build the set of neighborhood nodes is controlled from
the local reinit() method instead of just ensuring a second pass.

Also regold a test that yielded small differences after the update in this
commit. This test did not exactly match the 'always' patch update strategy results.
This latter approach is supposed to be more exact than iteration, so I take it
as the reference. After the change in this commit (i.e. building neighborhood lists
in every call to reinit instead of on construction) this regolded test that uses
'iteration' strategy yields the same results as the 'always' strategy. It looks as
if something got fixed by doing this.

GiudGiud commented:
Thanks for the patch!
